pg_bm25/src/telemetry/posthog.rs

```diff
@@ -0,0 +1,70 @@
+use pgrx::*;
+use serde_json::json;
+use std::env;
+use std::fs;
+
+pub unsafe fn init() {
+    // Read whether telemetry was already handled at the ParadeDB level
+    let telemetry_handled = match fs::read_to_string("/tmp/telemetry") {
+        Ok(content) => content.trim().to_string(),
+        Err(_) => String::from("false"),
+    };
+
+    // If telemetry was not already handled at the ParadeDB level, this is a standalone deployment
+    // of pg_bm25, and we should send telemetry
+    if telemetry_handled != "true" {
+        // We only send telemetry in the standalone deployment if TELEMETRY is set to true
+        let telemetry = env::var("TELEMETRY").unwrap_or_else(|_| String::from("false"));
+        if telemetry == "true" {
+            if let Ok(api_key) = env::var("POSTHOG_API_KEY") {
+                if let Ok(posthog_host) = env::var("POSTHOG_HOST") {
+                    if let Ok(commit_sha) = env::var("COMMIT_SHA") {
+                        // The endpoint for sending events to PostHog
+                        let endpoint: String = format!("{}/capture", posthog_host);
```

Comment:
I'm not sure if you see us needing to fiddle with this posthog configuration much, but I think it would be easier to extend if we restructure it to avoid the nested `if` blocks. I recommend the `envy` library for a `serde`-based approached to parsing the env-vars, and then a custom error type to more cleaning map the error messages:


```rust
use pgrx::*;
use serde::Deserialize;
use serde_json::json;
use std::fmt;

#[derive(Deserialize, Debug)]
struct Config {
    telemetry_handled: Option<String>, // Option because it's from a file, not env
    telemetry: String,
    posthog_api_key: String,
    posthog_host: String,
    commit_sha: String,
}

enum EnvError {
    PosthogApiKey,
    PosthogHost,
    CommitSha,
    ParsingError,
}

impl fmt::Display for EnvError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EnvError::PosthogApiKey => write!(f, "Failed to retrieve POSTHOG_API_KEY from environment variables, not sending pg_bm25 telemetry!"),
            EnvError::PosthogHost => write!(f, "Failed to retrieve POSTHOG_HOST from environment variables, not sending pg_bm25 telemetry!"),
            EnvError::CommitSha => write!(f, "Failed to retrieve COMMIT_SHA from environment variables, sending pg_bm25 telemetry without commit_sha!"),
            EnvError::ParsingError => write!(f, "Error parsing environment variables."),
        }
    }
}

impl Config {
    fn from_env() -> Result<Self, EnvError> {
        let telemetry_handled = std::fs::read_to_string("/tmp/telemetry")
            .map(|content| content.trim().to_string())
            .ok();

        let config = envy::from_env::<Config>().map_err(|_| EnvError::ParsingError)?;

        Ok(Config {
            telemetry_handled,
            ..config
        })
    }
}

pub unsafe fn init() {
    let config = Config::from_env();

    match config {
        Ok(config) => {
            if config.telemetry_handled.as_deref() == Some("true") || config.telemetry != "true" {
                return;
            }

            let endpoint = format!("{}/capture", config.posthog_host);
            let data = json!({
                "api_key": config.posthog_api_key,
                "event": "pg_bm25 Deployment",
                "distinct_id": uuid::Uuid::new_v4().to_string(),
                "properties": {
                    "commit_sha": config.commit_sha
                }
            });

            let client = reqwest::blocking::Client::new();
            let response = client
                .post(endpoint)
                .header("Content-Type", "application/json")
                .body(data.to_string())
                .send();

            match response {
                Ok(res) if res.status().is_success() => {
                    let _body = res.text().unwrap_or_else(|_| String::from("Failed to read response body"));
                }
                Ok(res) => {
                    info!("Failed to send event. Status: {}", res.status());
                }
                Err(e) => {
                    info!("Error sending request: {}", e);
                }
            }
        }
        Err(e) => {
            info!("{}", e);
        }
    }
}
```

Reply:
Oh wow, yeah this is nice, I'll do this